Build failure with Xcode 12 (clang 1200.0.32.2)

Building with Xcode 12 on macOS 10.15 leads to the following failure in `src/snprintf.c`:

```
2020-10-07T13:49:56.8826920Z src/snprintf.c:815:6: error: expected parameter declarator
2020-10-07T13:49:56.8827510Z  int vsnprintf (char *str, size_t count, const char *fmt, va_list args)
2020-10-07T13:49:56.8828190Z      ^
2020-10-07T13:49:56.8829450Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:75:40: note: expanded from macro 'vsnprintf'
2020-10-07T13:49:56.8830350Z   __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
2020-10-07T13:49:56.8830760Z                                        ^
2020-10-07T13:49:56.8831710Z src/snprintf.c:815:6: error: expected ')'
2020-10-07T13:49:56.8833330Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:75:40: note: expanded from macro 'vsnprintf'
2020-10-07T13:49:56.8834210Z   __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
2020-10-07T13:49:56.8834610Z                                        ^
2020-10-07T13:49:56.8835420Z src/snprintf.c:815:6: note: to match this '('
2020-10-07T13:49:56.8836710Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:75:29: note: expanded from macro 'vsnprintf'
2020-10-07T13:49:56.8838360Z   __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
2020-10-07T13:49:56.8839100Z                             ^
2020-10-07T13:49:56.8840890Z src/snprintf.c:815:6: error: conflicting types for '__builtin___vsnprintf_chk'
2020-10-07T13:49:56.8841960Z  int vsnprintf (char *str, size_t count, const char *fmt, va_list args)
2020-10-07T13:49:56.8842400Z      ^
2020-10-07T13:49:56.8843740Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:75:3: note: expanded from macro 'vsnprintf'
2020-10-07T13:49:56.8844620Z   __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
2020-10-07T13:49:56.8845010Z   ^
2020-10-07T13:49:56.8846220Z src/snprintf.c:815:6: note: '__builtin___vsnprintf_chk' is a builtin with type 'int (char *, unsigned long, int, unsigned long, const char *, struct __va_list_tag *)'
2020-10-07T13:49:56.8847860Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:75:3: note: expanded from macro 'vsnprintf'
2020-10-07T13:49:56.8848740Z   __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
2020-10-07T13:49:56.8849140Z   ^
2020-10-07T13:49:56.8850100Z src/snprintf.c:815:6: error: definition of builtin function '__builtin___vsnprintf_chk'
2020-10-07T13:49:56.8850790Z  int vsnprintf (char *str, size_t count, const char *fmt, va_list args)
2020-10-07T13:49:56.8851190Z      ^
2020-10-07T13:49:56.8852600Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:75:3: note: expanded from macro 'vsnprintf'
2020-10-07T13:49:56.8853480Z   __builtin___vsnprintf_chk (str, len, 0, __darwin_obsz(str), format, ap)
2020-10-07T13:49:56.8853870Z   ^
2020-10-07T13:49:56.8854740Z src/snprintf.c:817:29: error: use of undeclared identifier 'fmt'
2020-10-07T13:49:56.8855240Z     return dopr(str, count, fmt, args);
2020-10-07T13:49:56.8855550Z                             ^
2020-10-07T13:49:56.8856400Z src/snprintf.c:817:34: error: use of undeclared identifier 'args'
2020-10-07T13:49:56.8856930Z     return dopr(str, count, fmt, args);
2020-10-07T13:49:56.8857240Z                                  ^
2020-10-07T13:49:56.8857920Z src/snprintf.c:830:6: error: expected parameter declarator
2020-10-07T13:49:56.8858550Z  int snprintf(char *str,size_t count,const char *fmt,...)
2020-10-07T13:49:56.8858920Z      ^
2020-10-07T13:49:56.8860170Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:57:39: note: expanded from macro 'snprintf'
2020-10-07T13:49:56.8861030Z   __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
2020-10-07T13:49:56.8861390Z                                       ^
2020-10-07T13:49:56.8862440Z src/snprintf.c:830:6: error: expected ')'
2020-10-07T13:49:56.8863750Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:57:39: note: expanded from macro 'snprintf'
2020-10-07T13:49:56.8864600Z   __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
2020-10-07T13:49:56.8865510Z                                       ^
2020-10-07T13:49:56.8866440Z src/snprintf.c:830:6: note: to match this '('
2020-10-07T13:49:56.8867920Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:57:28: note: expanded from macro 'snprintf'
2020-10-07T13:49:56.8868770Z   __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
2020-10-07T13:49:56.8869150Z                            ^
2020-10-07T13:49:56.8870050Z src/snprintf.c:830:6: error: conflicting types for '__builtin___snprintf_chk'
2020-10-07T13:49:56.8870690Z  int snprintf(char *str,size_t count,const char *fmt,...)
2020-10-07T13:49:56.8871070Z      ^
2020-10-07T13:49:56.8872490Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:57:3: note: expanded from macro 'snprintf'
2020-10-07T13:49:56.8873340Z   __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
2020-10-07T13:49:56.8873720Z   ^
2020-10-07T13:49:56.8874860Z src/snprintf.c:830:6: note: '__builtin___snprintf_chk' is a builtin with type 'int (char *, unsigned long, int, unsigned long, const char *, ...)'
2020-10-07T13:49:56.8876440Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:57:3: note: expanded from macro 'snprintf'
2020-10-07T13:49:56.8877290Z   __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
2020-10-07T13:49:56.8877670Z   ^
2020-10-07T13:49:56.8878590Z src/snprintf.c:830:6: error: definition of builtin function '__builtin___snprintf_chk'
2020-10-07T13:49:56.8879260Z  int snprintf(char *str,size_t count,const char *fmt,...)
2020-10-07T13:49:56.8879640Z      ^
2020-10-07T13:49:56.8880930Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/secure/_stdio.h:57:3: note: expanded from macro 'snprintf'
2020-10-07T13:49:56.8882010Z   __builtin___snprintf_chk (str, len, 0, __darwin_obsz(str), __VA_ARGS__)
2020-10-07T13:49:56.8882390Z   ^
2020-10-07T13:49:56.8883330Z src/snprintf.c:835:18: error: use of undeclared identifier 'fmt'
2020-10-07T13:49:56.8883800Z     va_start(ap, fmt);
2020-10-07T13:49:56.8884050Z                  ^
2020-10-07T13:49:56.8887190Z clang -DMINILZO_HAVE_CONFIG_H -DHAVE_CONFIG_H -D_GNU_SOURCE -DLIBDIR="\"/usr/local/Cellar/distcc/3.3.3_3/lib\"" -DSYSCONFDIR="\"/usr/local/Cellar/distcc/3.3.3_3/etc\"" -DPKGDATADIR="\"/usr/local/Cellar/distcc/3.3.3_3/share/distcc\"" -Isrc -I"./src" -I"./lzo" -I"./popt" -Werror -I/private/tmp/distcc-20201007-10856-1qqh2j4/distcc-3.3.3/include -MD -W -Wall -Wimplicit -Wshadow -Wpointer-arith -Wcast-align -Wwrite-strings -Waggregate-return -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wmissing-declarations -D_THREAD_SAFE  -o src/loadfile.o -c src/loadfile.c
2020-10-07T13:49:56.8890020Z src/snprintf.c:836:33: error: use of undeclared identifier 'fmt'
2020-10-07T13:49:56.8890540Z     ret = vsnprintf(str, count, fmt, ap);
2020-10-07T13:49:56.8890860Z                                 ^
2020-10-07T13:49:56.8891140Z 12 errors generated.
2020-10-07T13:49:56.8891660Z make: *** [src/snprintf.o] Error 1
```

--- a/configure.ac
+++ b/configure.ac
@@ -433,9 +433,12 @@ fi
 
 AC_CACHE_CHECK([for C99 vsnprintf],rsync_cv_HAVE_C99_VSNPRINTF,[
 AC_TRY_RUN([
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
 #include <sys/types.h>
 #include <stdarg.h>
-void foo(const char *format, ...) {
+static void foo(const char *format, ...) {
        va_list ap;
        int len;
        char buf[5];
@@ -449,7 +452,7 @@ void foo(const char *format, ...) {
 
        exit(0);
 }
-main() { foo("hello"); }
+int main() { foo("hello"); }
 ],
 rsync_cv_HAVE_C99_VSNPRINTF=yes,rsync_cv_HAVE_C99_VSNPRINTF=no,rsync_cv_HAVE_C99_VSNPRINTF=cross)])
 if test x"$rsync_cv_HAVE_C99_VSNPRINTF" = x"yes"; then

--- a/configure.ac
+++ b/configure.ac
@@ -62,14 +62,13 @@ getpeername(0,0,&len);
 dnl Android requires -llog for printf and friends
 dnl and will only run position independent executables
 AC_MSG_CHECKING([if building on Android])
-AC_TRY_COMPILE([
+AC_TRY_COMPILE([], [
   #ifdef __BIONIC__
   int ok;
   (void)ok;
   #else
   choke me
   #endif],
-[func("a"); func("a", "b"); func("a", "b", "c")],
 [AC_MSG_RESULT(yes)
  LDFLAGS="$LDFLAGS -llog -pie"],
 [AC_MSG_RESULT(no)])
@@ -399,12 +398,11 @@ if (mmap (NULL, 0, 0, 0, 0, 0) == MAP_FAILED)
 	return 0;
 #else
 #error mmap unavailable
-#endif], ,[
+#endif], AC_MSG_RESULT(yes), [
+	AC_MSG_RESULT(no)
 	AC_DEFINE([MAP_FAILED], [(void *)-1L],
 	[Define if MAP_FAILED constant not available])
 ])
-AC_MSG_RESULT()
-
 
 dnl
 dnl Test if the preprocessor understand vararg macros
@@ -438,7 +436,7 @@ AC_TRY_RUN([
 #include <string.h>
 #include <sys/types.h>
 #include <stdarg.h>
-static void foo(const char *format, ...) {
+static int foo(const char *format, ...) {
        va_list ap;
        int len;
        char buf[5];
@@ -446,13 +444,13 @@ static void foo(const char *format, ...) {
        va_start(ap, format);
        len = vsnprintf(0, 0, format, ap);
        va_end(ap);
-       if (len != 5) exit(1);
+       if (len != 5) return 1;
 
-       if (snprintf(buf, 3, "hello") != 5 || strcmp(buf, "he") != 0) exit(1);
+       if (snprintf(buf, 3, "hello") != 5 || strcmp(buf, "he") != 0) return 1;
 
-       exit(0);
+       return 0;
 }
-int main() { foo("hello"); }
+int main() { return foo("hello"); }
 ],
 rsync_cv_HAVE_C99_VSNPRINTF=yes,rsync_cv_HAVE_C99_VSNPRINTF=no,rsync_cv_HAVE_C99_VSNPRINTF=cross)])
 if test x"$rsync_cv_HAVE_C99_VSNPRINTF" = x"yes"; then
@@ -465,9 +463,9 @@ AC_TRY_RUN([
 #include <sys/types.h>
 #include <sys/socket.h>
 
-main() {
+int main() {
        int fd[2];
-       exit((socketpair(AF_UNIX, SOCK_STREAM, 0, fd) != -1) ? 0 : 1);
+       return (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) != -1) ? 0 : 1;
 }],
 rsync_cv_HAVE_SOCKETPAIR=yes,rsync_cv_HAVE_SOCKETPAIR=no,rsync_cv_HAVE_SOCKETPAIR=cross)])
 if test x"$rsync_cv_HAVE_SOCKETPAIR" = x"yes"; then

